### PR TITLE
Cache version checking to speed up runs

### PR DESCRIPTION
**Pull request by mschwager**
This sped up subsequent `semgrep` runs quite a bit:

```
$ time python -m semgrep --validate --config ~/r2c/sgrep-rules/contrib/dlint/dlint-equivalent.yaml
Configuration is valid - found 1 valid configuration(s) and 0 configuration error(s).

real	0m1.031s
user	0m0.438s
sys	0m0.024s
```

```
$ time python -m semgrep --validate --config ~/r2c/sgrep-rules/contrib/dlint/dlint-equivalent.yaml
Configuration is valid - found 1 valid configuration(s) and 0 configuration error(s).

real	0m0.387s
user	0m0.366s
sys	0m0.022s
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -149,5 +149,4 @@ dmypy.json
 *.DS_Store
 
 # Ignore autogenerated version file
-semgrep/semgrep/version.py
```
Comment (mschwager) [inline .gitignore:152]: IIRC we're not autogenerating this file any more.

File: semgrep/semgrep/version.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+)
+VERSION_CHECK_TIMEOUT = int(
+    os.environ.get(
+        "SEMGREP_VERSION_CHECK_TIMEOUT", 2  # Don't block user's for too long
+    )
+)
+VERSION_CACHE_PATH = Path(
+    os.environ.get(
+        "SEMGREP_VERSION_CACHE_PATH",
+        Path(os.path.expanduser("~/.cache")) / "semgrep_version",
```
Comment (mschwager) [inline semgrep/semgrep/version.py]: Using `.cache` isn't ideal on OSX, but I figured this was fine compared to pulling in another dependency like [`appdirs`](https://pypi.org/project/appdirs/) just to grab a directory. If a user has strong feels about this they can always set `SEMGREP_VERSION_CACHE_PATH`.

**Comment by mschwager** [comment]
I was able to ape most of this from Bento, so it should be solid :+1:

File: semgrep/tests/unit/test_version.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+def test_version_check_caching(tmp_path):
+    tmp_cache_path = tmp_path / "semgrep_version"
+
+    time1 = time.time()
+    is_running_latest(tmp_cache_path)
+    time1 = time.time() - time1
+
+    time2 = time.time()
+    is_running_latest(tmp_cache_path)
```
Comment (brendongo) [inline semgrep/tests/unit/test_version.py]: If you want a more accurate test you can actually assert that _fetch_latest_version was only called once

Comment (brendongo) [reply semgrep/tests/unit/test_version.py]: Thoughts on copying the tests from https://github.com/returntocorp/bento/blob/master/tests/test_version.py ?

File: semgrep/semgrep/version.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+def _get_latest_version(version_cache_path: Path) -> Optional[str]:
+    latest_version_str = _get_version_from_cache(version_cache_path)
+    if latest_version_str is None:
+        latest_version_str = _fetch_latest_version()
+        if latest_version_str is None:
+            # Request timed out or invalid
+            return None
+
+        if not version_cache_path.parent.is_dir():
+            version_cache_path.parent.mkdir(parents=True, exist_ok=True)
```
Comment (brendongo) [inline semgrep/semgrep/version.py]: Do we need to check if the parent exists if we call mkdir with exist_ok=True anyway?

**Review by brendongo — APPROVED** [review]
Some non blocking questions/suggestions above

File: semgrep/semgrep/version.py (hunk 4 above)
Comment (mschwager) [reply semgrep/semgrep/version.py]: Good call!

**Pull request merged**